Initial version - T14298

krnowak opened the pull request:
https://phabricator.endlessm.com/T14298

krnowak committed:
initial version

krnowak committed:
add README.md

krnowak committed:
more todos

krnowak committed:
fix todos

krnowak committed:
fix markdown

krnowak committed:
fill appdata and desktop files

krnowak committed:
Add another TODO

krnowak committed:
Add a script for building the flatpak

krnowak committed:
some build fixes

krnowak committed:
finish flags

krnowak committed:
attempt to fix reading/writing from read-only dir

krnowak committed:
update .gitignore

krnowak committed:
update README

krnowak committed:
update the patch

krnowak committed:
Update git repos to more recent versions

The recent commits fix some compilation errors, so the build process
output is not that horrible anymore.

krnowak committed:
Update patches and patching process

Instead of enumerating the patches to be applied we use the wildcard
and assume that it will expand to the sorted list of git patches.

This also adds a patch for disabling the updates. This is a shot in
the darkness to fix the application hang at the startup.

krnowak committed:
Add another patch for forwarding env to the server

krnowak committed:
Reorganize superpowers patches

So far we had patches only for the app part of Superpowers, but there
will be patches for the other parts coming (for example for the core
part). So to avoid confusion which patches should be applied to which
part of Superpowers, put the existing patches for the app part in the
app-patches subdirectory.

krnowak committed:
Add a patch for core

krnowak committed:
Update patches

krnowak committed:
Update a patch

cosimoc commented on README.md:
```diff
@@ -0,0 +1,13 @@
+
+# TODO
+
+- currently we copy the core and app directory directly to `${prefix}` and remove the `.git` directories from there; probably we could copy even less files ([install script](https://github.com/kinvolk/superpowers-flatpak/blob/master/superpowers/install#L8-L11))
```
Maybe change the link to refer to the endless repo?

krnowak replied:
Fixed.

cosimoc commented on superpowers/app-patches/0002-Disable-updates.patch:
```diff
@@ -0,0 +1,30 @@
+Date: Mon, 16 Jan 2017 21:47:49 +0100
+Subject: [PATCH 2/3] Disable updates
+
+We don't want them inside the flatpak sandbox.
```
Perhaps the application could try to detect if it's running under flatpak and disable the updates at runtime in that case? Or you could add a configuration parameter/env var that flatpak would toggle.

krnowak replied:
Good idea, will do it.

lilic commented on build.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/bash
```
Not sure we need this script in the repo, maybe you can add it to the readme instead. Just a suggestion...

krnowak replied:
I would keep it, saves me a typing of the `flatpak-builder` command.

lilic commented on metadata/com.sparklinlabs.Superpowers.appdata.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.sparklinlabs.Superpowers.desktop</id>
```
you are missing a `<name>` field. See https://phabricator.endlessm.com/T15178#326355

krnowak replied:
Hm, seems I forgot to run the appstream validator here. Thanks for spotting.

krnowak replied:
I actually ran `appstream-util validate-relax resources/com.sparklinlabs.Superpowers.appdata.xml` and it is it's OK. I didn't run neither validate nor validate-strict, because they are way too picky.

lilic replied:
Yes, I know its valid, ran the same thing myself on other flatpaks, which ended up missing the `name` field. Its a bit odd.

lilic commented on com.sparklinlabs.Superpowers.json:
```diff
@@ -0,0 +1,143 @@
+                    "path": "superpowers/run-superpowers"
+                },
+                {
+                    "type": "file",
```
Any reason you aren't using type `patch` here?

krnowak replied:
IIRC, `patch` means that it will be copied to the toplevel directory of the module and then it will be applied there. I need to put it elsewhere to avoid mixing patches for different parts of the superpowers project and I need to enter a different directory to apply it.

lilic commented on com.sparklinlabs.Superpowers.json:
```diff
@@ -0,0 +1,143 @@
+            ]
+        },
+        {
+            "name": "metadata",
```
Due to flatpaks own metada, maybe call this `resources`, or something similar.

krnowak replied:
Alright.

lilic reviewed (changes requested):
Haven't tried building yet, will let you know the result later.

krnowak committed:
rename metadata module to resources

krnowak committed:
add the name tag to the appdata file

krnowak committed:
update README

lilic commented:
It builds and starts successfully, but this is the error I get:
`
Could not download 127.0.0.1:4237/superpowers.json
`

Another thing I noticed is that asset upload and download doesn't seem open the file selector.

krnowak committed:
point to the upstream repository, not to a fork

mattdangerw commented:
So I think at a high level we might be installing too much stuff here. Take a look at
https://github.com/superpowers/superpowers-app/blob/master/scripts/package.js
which will make the zip file they distribute for linux at the moment.

That will call into a tool called `electron-packager` which carries a lot of logic to make sure the released files only have whats needed, and often sticks assets and the like into an [asar](https://github.com/electron/asar) file.

I think it would be wise to either call into the package.js script directly, or make similar calls to `electron-packager` in our build flow. Don't see that happening right now.

krnowak committed:
update a patch disabling the updates

Now it disables them only when running inside the flatpak sandbox.

krnowak commented:
@mattdangerw: Yeah, I'm not calling anything from package.js other than a build script. For installing, I basically copy the stuff over to `/app/Superpowers` - https://github.com/endlessm/superpowers-flatpak/pull/1/files#diff-2276f4b83d2ce57fa62f710765f25113R8

mattdangerw commented:
Yup, I think it would be wiser to adapt the packaging flow that's already in place, which should leave you with a "distribution ready" directory of contents generated by `electron-packager` and copy that into the flatpak. There's a lot of logic in electron-packager I believe, though I certainly not super familiar with it, which it does not make sense to try to replicate here.

krnowak commented:
@mattdangerw: I will have to look at it more - note that we build two projects here - `app` and `core`. So I guess I will simply have to call `npm run package` under both `app` and `core` directories and then unpack the resulting archives in proper places.

krnowak committed:
update patches

krnowak committed:
install packages instead of everything

After building we do "npm run package" in app, and "npm run package"
and "npm run package game" in core. These commands generate the
packages ready to be distributed, so we copy their contents to the
specific locations. Hopefully this will make the flatpak a bit smaller.

krnowak committed:
remove unneeded electron-prebuilt module

It is downloaded when running "npm run package" in the app.

krnowak committed:
update deps file

krnowak committed:
update patches

krnowak committed:
update patches

krnowak committed:
update TODO

krnowak committed:
update patches

lilic commented on resources/com.sparklinlabs.Superpowers.appdata.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.sparklinlabs.Superpowers.desktop</id>
+  <name>Snap4Arduino</name>
```
s/Snap4Arduino/Superpowers

krnowak committed:
fix app name in the appdata file

krnowak commented:
Updated.

lilic reviewed (approved):
overall lgtm 👍
